core/src/main/java/org/apache/iceberg/hadoop/HadoopFileIO.java

```diff
@@ -84,4 +83,9 @@ public void setConf(Configuration conf) {
   public Configuration getConf() {
     return hadoopConf.get();
   }
+
+  @Override
+  public void setConfSupplier(SerializableSupplier<Configuration> confSupplier) {
```

Comment:
What about changing this to `Function<Configuration, SerializableSupplier<Configuration>>` so that the caller doesn't need to call `getConf`? It seems brittle to me that the caller is expected to set the supplier using the same conf that was taken from the `HadoopConfigurable`.